pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java

```diff
@@ -294,84 +296,99 @@ public RuleSetBuilder addRuleSetByReference(final RuleSet ruleSet, final boolean
         }
 
         /**
-         * Adds a new file exclusion pattern.
+         * Adds a new file exclusion pattern. The pattern must be a valid
+         * regular expression. If it is not, the pattern is ignored and the
+         * error is logged. With 7.0.0 we'll throw an exception instead.
          *
-         * @param aPattern
-         *            the pattern
-         * @return The same builder, for a fluid programming interface
+         * @param aPattern A valid regular expression
          */
         public RuleSetBuilder addExcludePattern(final String aPattern) {
-            if (!excludePatterns.contains(aPattern)) {
-                excludePatterns.add(aPattern);
+            try {
+                Pattern.compile(aPattern);
```

Comment:
this implies knowledge that:
1. These strings will be used only for `RegexStringFilter` instances
2. What the `RegexStringFilter` does within it's implementation (given in the name, true, but still implementation)

I'd either have `RegexStringFilter` complain itself inside the `optimize` method, or I'd cease to work with `Set<String>` here in favor of `Set<Pattern>`, and have `RegexStringFilter` receive a compiled `Pattern` instead of compiling itself (the "optimization" can still be done, as you can call `Pattern.pattern()`)

Reply:
Hum you're right, making those methods accept `Pattern` would be the easiest solution.  That way, we can complain about invalid regexes directly in the ruleset parser (closer to the input), which is where the string to Pattern conversion should happen, and other code should handle it.

So basically, deprecate all those methods and introduce new overloads which take Pattern instead?

Reply:
I used patterns directly instead. I don't want to touch RegexStringFilter, because for 7.0.0 I want to replace Filter with a standard Predicate, and RegexStringFilter can go away since its "optimisation" is completely insignificant, in view of profiling results.